### PR TITLE
Mjbots interface composition for pr

### PR DESCRIPTION
**Pull request by jdcaporale**
Most notably separated a RobotInterface class and the MjbotsRobotInterface class (whose name should likely change, perhaps to MjbotsInterface) in the control loop. This allowed the control loop to more easily take a RobotInterfaceDerived class as a template argument. Need to make some decisions on the names of function members for the control loop. Changes to existing examples and code should be minor, but this will certainly move some of the member names around.

### CODE REVIEW

File: examples/robot_example.cpp [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+#include "examples/simple_robot.h"
+
+class SimpleRobotControlLoop : public kodlab::mjbots::MjbotsControlLoop<ManyMotorLog, ModeInput, SimpleRobot>
+{
+    using MjbotsControlLoop::MjbotsControlLoop;
+
+    void CalcTorques() override
+    {
+        robot_->Update();
```
Comment (jdcaporale) [inline examples/robot_example.cpp:22]: Not sure if this is the interaction we want or if this should be hidden in robot unless you want an override

File: examples/robot_example.cpp [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        std::cout << "Switching to behavior " << robot_->mode << std::endl;
+        // If the kill robot mode is detected kill robot using CTRL_C flag handler.
+        if (robot_->mode == robot_->KILL_ROBOT)
+        { // KILL_ROBOT
+            kodlab::CTRL_C_DETECTED = true;
+        }
+    }
+};
+
+template <class joint_type>
```
Comment (jdcaporale) [inline examples/robot_example.cpp:54]: This class template could be moved to the joint base class probably though templating would get weird. Useful way to used shared pointers without forcing c++syntax. Probably a few other functions that could be moved in here like the getters and stuff, but right now this is implicitly converted to a vector of shared pointers.

Comment (jdcaporale) [reply examples/robot_example.cpp:54]: I thought about deriving std::vector but decided it was a bad idea with lots of ways it could go wrong.

File: examples/robot_example.cpp [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+int main(int argc, char **argv)
+{
+
+    // Setup joints
+    // std::vector<kodlab::mjbots::JointMoteus> joints;
+    // joints.emplace_back(100, 4, 1, 0,   1, 0);
+    // joints.emplace_back(101, 4,-1, 0, 5.0/3.0, 0);
+
+    JointSharedVector<kodlab::mjbots::JointMoteus> joints;
```
Comment (jdcaporale) [inline examples/robot_example.cpp:77]: Some of the simplifications that happened in the later commits mean this new class while useful as mentioned above may no longer be necessary.

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h [hunk 4]
```diff
@@ -42,14 +44,24 @@ struct ControlLoopOptions {
...
    */
   MjbotsControlLoop(std::vector<kodlab::mjbots::JointMoteus> joints, const ControlLoopOptions &options);
+  /*!
+   * \overload 
+   */
+  MjbotsControlLoop(std::vector<std::shared_ptr<kodlab::mjbots::JointMoteus>> joints, const ControlLoopOptions &options);
+  /*!
+   * \overload 
+   */
+  MjbotsControlLoop(std::shared_ptr<RobotClass>robot_in, const ControlLoopOptions &options);
```
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/mjbots_control_loop.h:64]: Needs comment for robot input

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h [hunk 5]
```diff
@@ -61,7 +73,7 @@ class MjbotsControlLoop : public AbstractRealtimeObject {
   /*!
    * @brief function to be implemented by child. Must set torques in the robot class
    */
-  virtual void CalcTorques() = 0;
+  virtual void CalcTorques() {robot_->Update();}
```
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/mjbots_control_loop.h:76]: this interaction could change (as mentioned above)

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h [hunk 6]
```diff
@@ -92,9 +104,15 @@ class MjbotsControlLoop : public AbstractRealtimeObject {
...
 
-  std::shared_ptr<MjbotsRobotInterface>
-      robot_;   /// ptr to the robot object, if unique causes many issues, also should be
-  /// initialized inside thread
+  /*!
+   * @brief Construct a new Setup Options object
+   * 
+   * @param options 
+   */
+  void SetupOptions(const ControlLoopOptions &options);
```
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/mjbots_control_loop.h:112]: useful for writing new constructors

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h [hunk 7]
```diff
@@ -109,16 +127,48 @@ class MjbotsControlLoop : public AbstractRealtimeObject {
...
+  SetupOptions(options);
+}
+
+template<class log_type, class input_type, class robot_type>
+MjbotsControlLoop<log_type, input_type, robot_type>::MjbotsControlLoop(std::shared_ptr<robot_type>robot_in, const ControlLoopOptions &options)
+  : AbstractRealtimeObject(options.realtime_params.main_rtp, options.realtime_params.can_cpu),
     lcm_sub_(options.realtime_params.lcm_rtp, options.realtime_params.lcm_cpu, options.input_channel_name) {
+  // Create robot object
+  robot_ = robot_in;
+  mjbots_interface_ = std::make_shared<kodlab::mjbots::MjbotsRobotInterface>( robot_->joints,
```
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/mjbots_control_loop.h:157]: This composition one of the important changes along with the additions of a RobotInterface class

File: lcm_types/mode_input.lcm [hunk 8]
```diff
@@ -0,0 +1,4 @@
+struct ModeInput
+{
+  int16_t    mode;
```
Comment (jdcaporale) [inline lcm_types/mode_input.lcm:3]: Should probably be a different PR?

File: src/mjbots_robot_interface.cpp [hunk 9]
```diff
@@ -47,28 +47,19 @@ ::mjbots::moteus::QueryResult MjbotsRobotInterface::Get(const std::vector<::mjbo
...
   for( auto & j: joints){
-    positions_.push_back( j->get_position_reference() );
-    velocities_.push_back( j->get_velocity_reference() );
-    torque_cmd_.push_back( j->get_servo_torque_reference()   ); 
     modes_.push_back(j->get_mode_reference());
   }
-  num_servos_ = joints.size();
-  for (size_t i = 0; i < num_servos_; ++i)
+  
+  for (size_t i = 0; i < num_joints_; ++i)
```
Comment (jdcaporale) [inline src/mjbots_robot_interface.cpp:62]: Changed for consistency still num_motors in controlLoop, should be the same everywhere?

File: lcm_types/mode_input.lcm (hunk 8 above)
Comment (ethanmusser) [reply lcm_types/mode_input.lcm:3]: Agreed, should probably rebase a81f6688a616721fc89d664e42eecd35c83525ee off of fd4b851c454f5986c7b01015873a63c63686b3dd

File: src/robot_interface.cpp [hunk 10]
```diff
@@ -0,0 +1,52 @@
+/*!
+ * @file robot_interface.cpp
```
Comment (ethanmusser) [inline src/robot_interface.cpp:2]: This class is mostly accessors and mutators.  Consider consolidating these portions of the implementation into `inline` functions in the class definition in `robot_interface.h`.

File: src/robot_interface.cpp [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  std::vector<float>torques(torque_cmd_.begin(), torque_cmd_.end());
+  return torques;
+}
+void kodlab::RobotInterface::SetTorques(std::vector<float> torques) {
+  soft_start_.ConstrainTorques(torques, cycle_count_);
+  float torque_cmd;
+  for (int joint_ind = 0; joint_ind < num_joints_; joint_ind++) {
+    torque_cmd = joints[joint_ind]->UpdateTorque(torques[joint_ind]);
+  }
+}
```
Comment (ethanmusser) [inline src/robot_interface.cpp:51]: This class is mostly accessors and mutators.  Consider consolidating these portions of the implementation into `inline` functions in the class definition in `robot_interface.h`.

File: include/examples/simple_robot.h [hunk 12]
```diff
@@ -0,0 +1,14 @@
+#include "kodlab_mjbots_sdk/robot_interface.h"
```
Comment (ethanmusser) [inline include/examples/simple_robot.h:1]: Missing file header for Doxygen

File: examples/robot_example.cpp (hunk 1 above)
Comment (ethanmusser) [reply examples/robot_example.cpp:29]: I think the latter.  This would permit the override if desired (and support backwards compatability) while encouraging use of custom `Robot` classes, which will provide better framework and added functionality (e.g. multiple behaviors).

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h (hunk 5 above)
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/mjbots_control_loop.h:84]: I think you should keep this default implementation (see other discussion in `robot_example.cpp`).

File: examples/robot_example.cpp (hunk 2 above)
Comment (ethanmusser) [reply examples/robot_example.cpp:66]: This should definitely move out of `robot_example`.  I don't think moving to the `JointBase` class makes sense as it is a vector composing one or several `JointBase`'s.  I would just keep this as-is, but stick it in its own class (perhaps in the `joint_base.h` header but outside of the `JointBase` class) to be included as needed.  The implicit conversion and wrapping of `std::shared_ptr` is nice for user interaction.

File: examples/robot_example.cpp (hunk 3 above)
Comment (ethanmusser) [reply examples/robot_example.cpp:102]: If it still provides some useful functionality, just stick it in its own header file to be included as needed.

File: include/kodlab_mjbots_sdk/common_header.h [hunk 13]
```diff
@@ -31,4 +31,20 @@ static void EnableCtrlC() {
   sigaction(SIGINT, &sigIntHandler, NULL);
   CTRL_C_DETECTED = false;
 }
+
+/**
+ * @brief Convert vector to shared pointers using copy constructor
+ * @param vect vector of type T
```
Comment (ethanmusser) [inline include/kodlab_mjbots_sdk/common_header.h:37]: Change `vect` to `objs`

File: src/mjbots_robot_interface.cpp (hunk 9 above)
Comment (ethanmusser) [reply src/mjbots_robot_interface.cpp:62]: Should be consistent everywhere, also `num_joints` is more apt.

File: examples/robot_example.cpp (hunk 2 above)
Comment (ethanmusser) [reply examples/robot_example.cpp:66]: I meant into the same header. My first thought was in something like the common_header but for joints and that is joint_base.h

Comment (jdcaporale) [reply examples/robot_example.cpp:66]: I meant into the same header. My first thought was in something like the common_header but for joints and that is joint_base.h

Comment (ethanmusser) [reply examples/robot_example.cpp:66]: Moved to joint_header

Comment (ethanmusser) [reply examples/robot_example.cpp:66]: Moved to joint_header

**Comment by ethanmusser** [comment]
### Thoughts

- Updated structure is more intuitive.  Using user-defined, `RobotInterface`-derived `Robot` classes is more easily expandable to complex robots with many composed classes and behaviors.  
- See comments in various discussions above.
- Improvements to #16 require the updated infrastructure.  We should get #22 (or both #16 and #22) merged to make these changes.

### TODOs

- [ ] Move LCM mode input addition to separate commit (i.e. merge a separate PR for mode input, then rebase fd4b851c454f5986c7b01015873a63c63686b3dd off of the new commit and rebase a81f6688a616721fc89d664e42eecd35c83525ee off of fd4b851c454f5986c7b01015873a63c63686b3dd)
- [ ] Consolidate `RobotInterface` implementation to `robot_interface.h` only
- [ ] Figure out `CalcTorques` interaction per discussions above
- [ ] Update all usages to `num_joints_` per discussions above
- [ ] General cleanup (i.e., Doxygen comments, format changes per Drake style guide)

File: src/robot_interface.cpp (hunk 11 above)
Comment (jdcaporale) [reply src/robot_interface.cpp:51]: This makes sense. I will do this unless told otherwise. @ShaneRozenLevy?

File: examples/robot_example.cpp [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+    operator std::vector<std::shared_ptr<joint_type>>() { return v_; }
+};
+
+int main(int argc, char **argv)
+{
+
+    // Setup joints with a std::vector
+    // std::vector<kodlab::mjbots::JointMoteus> joints;
+    // joints.emplace_back(100, 4, 1, 0,   1, 0);
+    // joints.emplace_back(101, 4,-1, 0, 5.0/3.0, 0);
```
[the remark this answers is not in the document]
Comment (jdcaporale) [reply examples/robot_example.cpp]: Not extraneous, mean to show that you can do it either way

File: examples/robot_example.cpp (hunk 3 above)
Comment (jdcaporale) [reply examples/robot_example.cpp]: Done. Also leaving both implementations in the example.

File: lcm_types/mode_input.lcm (hunk 8 above)
Comment (ethanmusser) [reply lcm_types/mode_input.lcm:3]: See #24

File: examples/robot_example.cpp [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+#include "examples/simple_robot.h"
+
+class SimpleRobotControlLoop : public kodlab::mjbots::MjbotsControlLoop<ManyMotorLog, ModeInput, SimpleRobot>
+{
+    using MjbotsControlLoop::MjbotsControlLoop;
+
+    void CalcTorques() override
+    {
+        robot_->Update();
```
Comment (ShaneRozenLevy) [inline examples/robot_example.cpp:29]: I would add some code here that actually uses the robot to set some torques or atleast a comment expressing that idea as a code pattern.

File: include/examples/simple_robot.h [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+{
+
+    using kodlab::RobotInterface::RobotInterface;
+
+public:
+    int mode = 0; // Member variables encode whatever added state we need
+    // Set up robot update function for state and torques
+    void Update() override
+    {
+        std::vector<float> torques(num_joints_, 0);
```
Comment (ShaneRozenLevy) [inline include/examples/simple_robot.h:24]: I don't think this is what we want in the update as an example for people to pattern match. Maybe give it a leg and use that??

File: include/kodlab_mjbots_sdk/robot_interface.h [hunk 17]
```diff
@@ -0,0 +1,143 @@
...
+#include "kodlab_mjbots_sdk/joint_base.h"
+#include "kodlab_mjbots_sdk/soft_start.h"
+
+namespace kodlab
+{
+    /*!
+     * @brief Sets up a kodlab robot interface class 
+     * 
+     */
+    class RobotInterface {
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/robot_interface.h:25]: we probably want it to be RobotBase rather than RobotInterface

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h (hunk 5 above)
Comment (ShaneRozenLevy) [reply include/kodlab_mjbots_sdk/mjbots_control_loop.h:84]: I think we should not change this. Keep it as: `  virtual void CalcTorques() = 0;`

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h [hunk 18]
```diff
@@ -92,33 +112,71 @@ class MjbotsControlLoop : public AbstractRealtimeObject {
...
-  /// initialized inside thread
+  /*!
+   * @brief Construct a new Setup Options object
+   * 
+   * @param options 
+   */
+  void SetupOptions(const ControlLoopOptions &options);
+
+  std::shared_ptr<RobotClass> robot_;     /// ptr to the robot object
+  std::shared_ptr<kodlab::mjbots::MjbotsRobotInterface> mjbots_interface_;   ///ptr to mjbots_interface object, if unique causes issues, also should be initialized inside thread
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/mjbots_control_loop.h:123]: Should this be a unique ptr?

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h [hunk 19]
```diff
@@ -189,25 +241,27 @@ void MjbotsControlLoop<log_type, input_type>::Run() {
...
+      mjbots_interface_->ProcessReply();
       prev_msg_duration = message_duration_timer.tac();
       message_duration_timer.tic();
-      robot_->SendCommand();
+      mjbots_interface_->SendCommand();
     }
+
     // Calculate torques and log
-    CalcTorques();
+    CalcTorques();      //TODO should we give full control to the robot_ instead of feeding update thorugh?
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/mjbots_control_loop.h:251]: Lets rename CalcTorques() to Update() and have Update be up to the user what it does. They can call the robot update if they want or just do vanilla calc torques.

File: include/kodlab_mjbots_sdk/mjbots_robot_interface.h [hunk 20]
```diff
@@ -30,32 +30,29 @@ struct RealtimeParams {
   int lcm_cpu = 0;
 };
 
-class MjbotsRobotInterface {
+/*!
+ * @brief A class that allows interaction with the Mjbots Moteus Motor Controllers
+ * 
+ */
+class MjbotsRobotInterface  {
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/mjbots_robot_interface.h:37]: Lets rename this class to `MjbotsHardwareInterface`

File: include/kodlab_mjbots_sdk/robot_interface.h [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    public:
+        static const int KILL_ROBOT = -1;
+        std::vector< std::shared_ptr<JointBase> > joints; ///the vector of shared_ptrs to joints
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/robot_interface.h:28]: create getter

File: include/kodlab_mjbots_sdk/robot_interface.h [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+            // Ensure at compile time that the template is JointBase or a child of JointBase
+            static_assert(std::is_base_of<JointBase, JointDerived>::value);
+
+            // Set up joint vector and joint states
+            for (std::shared_ptr<JointDerived> j : joint_vect){
+                joints.push_back(j); // Copy constructed, uses an implicit upcasting
+                positions_.push_back( j->get_position_reference() );
+                velocities_.push_back( j->get_velocity_reference() );
+                torque_cmd_.push_back( j->get_servo_torque_reference() ); 
+            }
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/robot_interface.h:52]: Maybe here create a map from joint name to joint_id.

File: include/kodlab_mjbots_sdk/robot_interface.h [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+        std::vector<float> GetJointTorqueCmd();
+
+        //TODO virtual Attitude GetAttitude()
+
+        /*! 
+         * @brief Get sub-vector of shared_ptr to joint objects via a set of indices
+         * @param joint_indices set of desired joint indices as std::vector of ints
+         * @return a vector shared pointers to the desired joints
+         */
+        std::vector<std::shared_ptr<JointBase>> GetJoints(std::vector<int> joint_indices);
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/robot_interface.h:110]: Index here by joint_ids rather than by ints.

File: include/kodlab_mjbots_sdk/robot_interface.h [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+        void SetTorques(std::vector<float> torques);
+
+        /*!
+         * @brief accessor for the torque md, takes into account direction
+         * @return the torque cmd
+         */
+        std::vector<float> GetJointTorqueCmd();
+
+        //TODO virtual Attitude GetAttitude()
+
```
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/robot_interface.h:104]: Create a find joint_id_by_name_fatal function

File: src/robot_interface.cpp (hunk 11 above)
Comment (ShaneRozenLevy) [reply src/robot_interface.cpp:51]: seems fine.

**Review by ShaneRozenLevy — COMMENTED** [review]
seems like a good start and a correct implementation of what we wanted to do. I think as we look towards making more PRs its important to break them into small pieces so they are easier to review.

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h (hunk 19 above)
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/mjbots_control_loop.h]: Renamed in 2d28dd28bc0957d92233c7ff3f997166d212ee69.

File: examples/robot_example.cpp (hunk 1 above)
Comment (ethanmusser) [reply examples/robot_example.cpp:29]: I think we should not change this. Keep it as: `  virtual void CalcTorques() = 0;`

_Originally posted by @ShaneRozenLevy in https://github.com/KodlabPenn/kodlab_mjbots_sdk/pull/22#discussion_r931545298_

File: src/mjbots_robot_interface.cpp (hunk 9 above)
Comment (ethanmusser) [reply src/mjbots_robot_interface.cpp:62]: Resolved in 8d4fd951d65b7cc23aedfa556c2617725146633e.

File: examples/robot_example.cpp (hunk 1 above)
Comment (ethanmusser) [reply examples/robot_example.cpp:29]: I think Shane has a point.  Keep this in `robot_example.cpp` and comment appropriately.  Then make `CalcTorques` (now `Update`) pure virtual in `MjbotsControlLoop`.

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h (hunk 5 above)
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/mjbots_control_loop.h]: I've come around to Shane's take so long as a `CalcTorques()` override is provided in `robot_example.cpp` that includes `robot_->Update()` and comments explaining its usage.

Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/mjbots_control_loop.h]: Resolved in 060a657fd0508b266f5dde0c99b0d6b2a72ed7e7.

File: examples/robot_example.cpp (hunk 1 above)
Comment (ethanmusser) [reply examples/robot_example.cpp:29]: `CalcTorqes` made pure virtual in 060a657fd0508b266f5dde0c99b0d6b2a72ed7e7. Documentation of this override should be clarified in a robot_example.cpp documentation update commit.

File: include/kodlab_mjbots_sdk/mjbots_robot_interface.h (hunk 20 above)
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/mjbots_robot_interface.h]: Renamed in ea7778a442081ea4ef04a3c9f17389b0c9217d31.

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h (hunk 7 above)
Comment (jdcaporale) [reply include/kodlab_mjbots_sdk/mjbots_control_loop.h]: Everyone seems on board with this.

File: examples/robot_example.cpp (hunk 1 above)
Comment (jdcaporale) [reply examples/robot_example.cpp:29]: LGTM

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h (hunk 18 above)
Comment (jdcaporale) [reply include/kodlab_mjbots_sdk/mjbots_control_loop.h]: Future plans share the robot interface, its possible that it is useful to be able to share the hardware interface as well, and shared pointers are easier to manage. 
Unique pointer would work for now and could just be passed as a raw pointer, but that is probably bad form.

File: include/kodlab_mjbots_sdk/mjbots_control_loop.h (hunk 19 above)
Comment (jdcaporale) [reply include/kodlab_mjbots_sdk/mjbots_control_loop.h]: Because of bug in soft_start this will likely break code, but will be fixed before merged.

- [ ] Fix soft start timer - get rid of cycle count

File: include/kodlab_mjbots_sdk/robot_interface.h (hunk 17 above)
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/robot_interface.h]: Renamed in 23bee8968f2a74e9649946cb098f472349cbd7ac.

File: include/kodlab_mjbots_sdk/robot_interface.h (hunk 21 above)
Comment (jdcaporale) [reply include/kodlab_mjbots_sdk/robot_interface.h]: Currently public, do we need a getter here as well?

Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/robot_interface.h]: I don't think we need a getter while this is public.  `joints` could be made private and have access limited to an accessor.

File: include/kodlab_mjbots_sdk/robot_interface.h (hunk 22 above)
Comment (jdcaporale) [reply include/kodlab_mjbots_sdk/robot_interface.h]: I think that is a good idea. I'm moving it to an issue (#28)  to be done in it's own PR or with the joint base name  PR (#27).

File: include/kodlab_mjbots_sdk/robot_interface.h (hunk 23 above)
Comment (jdcaporale) [reply include/kodlab_mjbots_sdk/robot_interface.h]: I think this should be handled by maps/typedef as well. In particular, so that we can get the joints by CAN_ID or INDEX with INDEX being of type Joint_Index_t or something like that. Related to #28.

File: include/kodlab_mjbots_sdk/robot_interface.h (hunk 24 above)
Comment (jdcaporale) [reply include/kodlab_mjbots_sdk/robot_interface.h]: What exactly do you mean by fatal? Should we handle the id error and CTRL-C? Allow the error to kill the program? Or some other jargon I don't know? 
Also, we will be making a map as documented in #28 and this will be handled there instead.

**Comment by jdcaporale** [comment]
Naming convention changes LGTM and new update too (last five commits)

File: include/examples/simple_robot.h (hunk 16 above)
Comment (jdcaporale) [reply include/examples/simple_robot.h:26]: Will add better example code

Comment (jdcaporale) [reply include/examples/simple_robot.h:26]: #36
